### DISCUSSION TITLE
Flexsearch isn't finding some of the pages

### DISCUSSION THREAD

**Discussion by myrthos** [gethinode/hinode#450] [Help]
Some of the pages in my site's documentation pages are not indexed by Flexsearch.  
For the documentation of shortcodes, the [overview](https://myrthos.net/docs/shortcodes/introduction/overview/) page is indexed, but all other pages are not.

Undoubtedly this is due to the different way that I handle documentation. One of the differences is that I use a page bundle file `_index.md` to list all the available documentation sets. As this would normally list all the pages in the documentation set. I have for all pages, except one, in the frontmatter:

```
_build:
  list: never
```

To prevent those pages to be listed. The one page that does not have this, is the entry page for the documentation set and is also the one that is properly indexed by Flexsearch. All other pages are not indexed.

Perhaps this is causing Flexsearch to not index those pages?

I looked in `flexsearch_index.js` and this is where the filtering happens:

```
{{ $list := where (where site.AllPages "Kind" "in" "page") "Title" "!=" "" }}
```

Using the build options [documentation](https://gohugo.io/content-management/build-options/) I assume `list: never` might cause the page to not be found by `AllPages`.

My question is if this indeed causes the pages to not be found by Flexsearch? And if so, what would be a potential solution?

**Comment by markdumay** [comment]
I suspect you're right. Have you considered adjusting the overview page instead of excluding all individual pages? 

Hinode uses configurable page sections. What happens if you set `nested` to false for this particular section (~documentation sets)?
See https://gethinode.com/docs/configuration/layout/#configuration-4.

**Comment by myrthos** [comment]
I woke up this morning with the idea to remove the build option from all the pages and modify `section-list.html` specifically for documentation sets to just show one page per set. But, if there is a solution that makes merging easier in the future that is preferable obviously.

In my case it would not make a difference if `nested` is true or false as there are no other `_index.md` pages below the one in `docs` and because of that there are no nested sections. I could change that by adding in each documentation set an `_index.md` page and setting `nested` to false in that page, but in that case I think it would just find nothing. The 'section-list.html' file uses `.RegularPages` and because `nested` is false, I think it finds no pages. By changing `.RegularPages` to `Pages` it would return the `_index.md` page in the documentation set but none of the regular pages, which would be much closer to what I need. Unless I completely misunderstand how this works, but I  made that change to 'section-list.html' already for supporting multi-page articles and it seems to work like that for that solution. Should not be too hard to test if it also works for my documentation sets.

If I am misinterpreting how this works, please let me know, and thank you for helping me finding a solution.

**Comment by markdumay** [comment]
Perhaps I'm misinterpreting your configuration/intention, so please forgive me. ;-) I introduced the `nested` argument to change the behavior of list pages. By default, Hugo collects all regular pages within a list page - this includes nested (~child) pages. The following tree illustrates this:

```yaml
- title: A # lists B, C, First, Second, Third
  pages:
    - title: B # lists C, First, Second, Third
      pages:
        - title: C # lists First, Second, Third
          pages:
            - title: First
            - title: Second
            - title: Third
```

Modifying `nested` would change that behavior:

```yaml
- title: A # lists B
  pages:
    - title: B # lists C
      pages:
        - title: C # lists First, Second, Third
          pages:
            - title: First
            - title: Second
            - title: Third
```

I don't think it matters if B or C contains an `_index.md` page, but I could be wrong.

**Comment by myrthos** [comment]
I tried this out with removing the build options and setting `nested` to false and to true. It makes no difference whether or not `nested` is false or true. I get all the document pages. I'm using a directory structure similar to this:

```
content
   docs
      _index.md
      docset1
         introduction
            intro.md
         part-a
            page1.md
            page2.md
```

If I now add to the `part-a` folder the file `_index.md`, it will look like this:

```
content
   docs
      _index.md
      docset1
         introduction
            intro.md
         part-a
            _index.md
            page1.md
            page2.md
```

Now there is a difference in behavior between `nested` being false or true (in `docs/_index.md`). If `nested` is true. I get all three pages as before.  
If `nested` is false, I only get the intro file and not any of the files in the `part-a` folder. This would correspond with what the [Hugo docs](https://gohugo.io/content-management/sections/#nested-sections) write about nested sections.  
This solution does result in all the pages becoming part of the search. The downside of it is that `/docs/part-a` has become an existing page (because of the `_index.md` file in that folder) and if someone would manually go to that page, it would show the pages in that folder (`nested` is false) or all pages in docs (`nested` is true).  
So I added this to the frontmatter of the `_index.md` file in part-a:

```yaml
---
_build:
    render: false
---
```

This makes the page no longer linkable, while the pages in that folder are still part of the search.

So, in principle using `nested: false` and adding `_index.md` files to the folders, with the exception of the folder with the landing page (in this case `introduction`), my issues are solved. There is just one downside to this. There can be only one page in the `introduction` folder. If there are more pages they are all listed. I don't think that can be solved without modifying `section-list.html`, so for now I'll just take that into account.

I hope this makes sense.

**Comment by myrthos** [comment]
I've implemented the above solution on my site and it works. In the morning I wanted to modify code, but because of your feedback, I can now implement it without modifying any code. So, thanks for leading me on that track.

For completeness (actually curiosity) I also tried to  overcome the remaining downside, which I listed in the previous post by modifying `section-list.html` and adding the following:

```
{{ if eq $page.Params.layout "docs" }}
    {{ $list = where $page.RegularPages "Params.landing" true }}
{{ end }}
```

I also needed to undo the changes from the previous post. After that the `$list` variable  only holds the pages that have `landing: true` in the frontmatter, which would be the case for only one page per documentation set.

My curiosity has been satisfied :smile:

**Comment by markdumay** [comment]
Great, it seems that you've found a workable solution! Any recommendations you'd like to make for the Hinode repo, or is the current approach adequate?

**Comment by myrthos** [comment]
No, I have no suggestions for the Hinode repo. The way I handle documentation is very specific for my needs and I don't think any of that should flow back into Hinode.
